Excluded paths in nested configurations are ignored

### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

My example project has this structure:

```
.
├── .swiftlint.yml
└── A
    ├── .swiftlint.yml
    ├── B
    │   └── Longcat.swift
    └── Hello.swift
```

`./swiftlint.yml` is empty, `./A/swiftlint.yml` has the following contents:

```yaml
excluded:
  - B
```

When I run `swiftlint` from the project's root directory, `Longcat.swift` gets linted although it should be excluded.

An explanation as to why I'm trying to do this: We have a common base `.swiftlint.yml` for our apps, which we symlink from a CocoaPod. Since we need to add project-specific excludes as well (and SwiftLint unfortunately doesn't seem to support shared base configurations at all), I need to mark individual folders as excluded from nested configurations.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths
Linting 'Longcat.swift' (1/2)
Linting 'Hello.swift' (2/2)
.../swiftlint-example/A/B/Longcat.swift:1: warning: Line Length Violation: Line should be 120 characters or less: currently 189 characters (line_length)
Done linting! Found 1 violation, 0 serious in 2 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?

`0.27.0`

* Installation method used (Homebrew, CocoaPods, building from source, etc)?

In this instance, I was using the `swiftlint` from `brew install swiftlint`

* Paste your configuration file:

See above.

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

Yes. Toplevel is empty, nested config is the one shown above.

* Which Xcode version are you using (check `xcode-select -p`)?

`Version 10.0 (10A255)`

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

I've attached my minimal example project: [swiftlint-nested-exclude-example.zip](https://github.com/realm/SwiftLint/files/2526054/swiftlint-nested-exclude-example.zip)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -49,8 +49,8 @@
 #### Bug Fixes
 
 * Fix bug where SwiftLint ignores excluded files list in a nested configuration file.  
-[Dylan Bruschi](https://github.com/Bruschidy54)
-[#2447](https://github.com/realm/SwiftLint/issues/2447)
+  [Dylan Bruschi](https://github.com/Bruschidy54)
+  [#2447](https://github.com/realm/SwiftLint/issues/2447)
 
 * Fix false positives on `no_grouping_extension` rule when using `where`
   clause.  

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -48,6 +48,10 @@
 
 #### Bug Fixes
 
+* Fix bug where SwiftLint ignores excluded files list in a nested configuration file.  
+[Dylan Bruschi](https://github.com/Bruschidy54)
+[#2447](https://github.com/realm/SwiftLint/issues/2447)
+
 * Fix false positives on `no_grouping_extension` rule when using `where`
   clause.  
   [Almaz Ibragimov](https://github.com/almazrafi)

--- a/Source/swiftlint/Extensions/Configuration+CommandLine.swift
+++ b/Source/swiftlint/Extensions/Configuration+CommandLine.swift
@@ -61,7 +61,34 @@ extension Configuration {
             let errorMessage = "No lintable files found at paths: '\(visitor.paths.joined(separator: ", "))'"
             return .failure(.usageError(description: errorMessage))
         }
-        return .success(Dictionary(grouping: files, by: configuration(for:)))
+
+        var groupedFiles = [Configuration: [File]]()
+        for file in files {
+            // Files whose configuration specifies they should be excluded will be skipped
+            let fileConfiguration = configuration(for: file)
+            let excludedPaths = fileConfiguration.excluded
+                .compactMap { fileConfiguration.rootPath?.bridge().appendingPathComponent($0) }
+
+
+            var shouldSkip = false
+            for excludedPath in excludedPaths {
+                if file.path?.bridge().pathComponents.starts(with: excludedPath.bridge().pathComponents) ?? false {
+                    shouldSkip = true
+                    break
+                }
+            }
+
+            if !shouldSkip {
+                if var configuredFiles = groupedFiles[fileConfiguration] {
+                    configuredFiles.append(file)
+                    groupedFiles[fileConfiguration] = configuredFiles
+                } else {
+                    groupedFiles[fileConfiguration] = [file]
+                }
+            }
+        }
+
+        return .success(groupedFiles)
     }
 
     private func visit(filesPerConfiguration: [Configuration: [File]],

--- a/Source/swiftlint/Extensions/Configuration+CommandLine.swift
+++ b/Source/swiftlint/Extensions/Configuration+CommandLine.swift
@@ -69,8 +69,8 @@ extension Configuration {
             let excludedPaths = fileConfiguration.excluded
                 .map { (fileConfiguration.rootPath ?? "").bridge().appendingPathComponent($0) }
 
-            let shouldSkip: Bool = excludedPaths.contains
-            { file.path?.bridge().pathComponents.starts(with: $0.bridge().pathComponents) ?? false }
+            let shouldSkip: Bool = excludedPaths.contains { file.path?.bridge().pathComponents
+                .starts(with: $0.bridge().pathComponents) ?? false }
 
             if !shouldSkip {
                 if var configuredFiles = groupedFiles[fileConfiguration] {

--- a/Source/swiftlint/Extensions/Configuration+CommandLine.swift
+++ b/Source/swiftlint/Extensions/Configuration+CommandLine.swift
@@ -69,7 +69,6 @@ extension Configuration {
             let excludedPaths = fileConfiguration.excluded
                 .compactMap { fileConfiguration.rootPath?.bridge().appendingPathComponent($0) }
 
-
             var shouldSkip = false
             for excludedPath in excludedPaths {
                 if file.path?.bridge().pathComponents.starts(with: excludedPath.bridge().pathComponents) ?? false {

--- a/Source/swiftlint/Extensions/Configuration+CommandLine.swift
+++ b/Source/swiftlint/Extensions/Configuration+CommandLine.swift
@@ -69,8 +69,9 @@ extension Configuration {
             let excludedPaths = fileConfiguration.excluded
                 .map { (fileConfiguration.rootPath ?? "").bridge().appendingPathComponent($0) }
 
-            let shouldSkip: Bool = excludedPaths.contains { file.path?.bridge().pathComponents
-                .starts(with: $0.bridge().pathComponents) ?? false }
+            let shouldSkip: Bool = excludedPaths.contains {
+                file.path?.bridge().pathComponents.starts(with: $0.bridge().pathComponents) ?? false
+            }
 
             if !shouldSkip {
                 if var configuredFiles = groupedFiles[fileConfiguration] {

--- a/Source/swiftlint/Extensions/Configuration+CommandLine.swift
+++ b/Source/swiftlint/Extensions/Configuration+CommandLine.swift
@@ -67,15 +67,10 @@ extension Configuration {
             // Files whose configuration specifies they should be excluded will be skipped
             let fileConfiguration = configuration(for: file)
             let excludedPaths = fileConfiguration.excluded
-                .compactMap { fileConfiguration.rootPath?.bridge().appendingPathComponent($0) }
+                .map { (fileConfiguration.rootPath ?? "").bridge().appendingPathComponent($0) }
 
-            var shouldSkip = false
-            for excludedPath in excludedPaths {
-                if file.path?.bridge().pathComponents.starts(with: excludedPath.bridge().pathComponents) ?? false {
-                    shouldSkip = true
-                    break
-                }
-            }
+            let shouldSkip: Bool = excludedPaths.contains
+            { file.path?.bridge().pathComponents.starts(with: $0.bridge().pathComponents) ?? false }
 
             if !shouldSkip {
                 if var configuredFiles = groupedFiles[fileConfiguration] {

